### ISSUE TITLE
Add hooks to do something before and after the drag started and ended

### ISSUE THREAD

**Issue by awaescher** [awaescher/FluentDragDrop#17]
Slightly related to #14 : The only clean option to update the UI once a drag&drop operation ended is using the `To()` extension that provides a Func to define what happens once the user finishes the drag&drop operation successfully.

```csharp
pic.InitializeDragAndDrop()
  .Copy()
  .OnMouseMove()
  .WithData(() => pic.Image)
  .To(PreviewBoxes, (target, data) =>
  {
    target.Image = data;
    MessageBox.Show("...");        //  <-----
  });
```

However, if the target controls are not known while initializing the drag&drop operation, `To()` cannot be used. And then, there's no clean way to react on the operation once it is finished.
In addition, there's no clean way to update the UI as soon as a (deferred) drag operation stars, except a workaround which is updating it within the lambda to define the data to drag:

```csharp
pic.InitializeDragAndDrop()
  .Copy()
  .OnMouseMove()
  .WithData(() => 
  {
    UpdateUi();        //  <-----
    return pic.Image;
  })
  .To(PreviewBoxes, (target, data) =>
  {
    target.Image = data;
    MessageBox.Show("...");
  });
```

Provide hooks (within the UI thread):
 - Start of the operation
 - End of the operation (bool wasCanceled)

**Proposal:**

```csharp
pic.InitializeDragAndDrop()
  .Copy()
  .OnMouseMove()
  .WithData(() => pic.Image)
  .To(PreviewBoxes, (target, data) => target.Image = data)
  .OnStart(() => UpdateUi())
  .OnEnd(success => MessageBox.Show(success ? "Done" : "Canceled"));
```

**Comment by awaescher** [comment]
Hooks are available with the next version and are guaranteed to be run on the UI thread:

```
pic.InitializeDragAndDrop()
  .Copy()
  .OnMouseMove()
  .WithData(() => pic.Image)
  .To(PreviewBoxes, (target, data) => target.Image = data)
  .OnBeforeStart(() => ...)
  .OnAfterStart(() => ...)
  .OnDrop(() => MessageBox.Show("Dropped successfully"))
  .OnCancel(() => MessageBox.Show("Canceled"));
```
